File: matlab/examples/CHRJ_3comp/example_HEFA.m
%     ____                  _____ ___    ______
%    / __ )____ ___  _____ / ___//   |  / ____/
%   / __  / __ `/ / / / _ \__ \/ /| | / /_
%  / /_/ / /_/ / /_/ /  __/__/ / ___ |/ __/
% /_____/\__,_\__, /\___/____/_/  |_/_/
%             /____/
%
% BayeSAF: Emulation and Design of Sustainable Alternative Fuels
% via Bayesian Inference and Descriptors-Based Machine Learning
%
% Contributors / Copyright Notice
% © 2026 Jacopo Liberatori — jacopo.liberatori@centralesupelec.fr
% Postdoctoral Researcher @ Laboratoire EM2C, CentraleSupélec (CNRS)
%
% © 2026 Davide Cavalieri — davide.cavalieri@uniroma1.it
% Postdoctoral Researcher @ Sapienza University of Rome,
% Department of Mechanical and Aerospace Engineering (DIMA)
%
% © 2026 Matteo Blandino, Ph.D.
%
% Reference:
% J. Liberatori, D. Cavalieri, M. Blandino, M. Valorani, and P.P. Ciottoli.
% BayeSAF: Emulation and Design of Sustainable Alternative Fuels via Bayesian
% Inference and Descriptors-Based Machine Learning. Fuel 419, 138835 (2026).
% Available at: https://doi.org/10.1016/j.fuel.2026.138835.

% DESCRIPTION:
% The present example illustrates how to develop a three-component surrogate
% mixture emulating the HEFA-SPK POSF-6152 (CHRJ) fuel.

clc
clear
close all

addpath('../../thermo-transport')
addpath('../../distillation')
addpath('../../pdfs')
addpath('../../mcmc')
addpath('../../postprocessing')
addpath('../../gsa')
addpath('../../utilities')

addpath('../../../database/nparaffins')
addpath('../../../database/isoparaffins')
addpath('../../../database/cycloparaffins')
addpath('../../../database/dicycloparaffins')
addpath('../../../database/alkylbenzenes')
addpath('../../../database/alkylnaphtalenes')
addpath('../../../database/cycloaromatics')

addpath('../../../exp/CHRJ_POSF6152')

%% --- REAL FUEL DATA  --- %%

% CSV files containing the experimental data about the thermophysical properties of the real fuel
Dataset = {'rho', 'nu', 'distillation', 'deltaT_dist', 'molWeight', 'HC', 'DCN', 'LHV', 'flash', 'freezing'};
% Plot legend string denoting real fuel experimental data
fuel_name = 'CHRJ POSF-6152';
% Bubble temperature of the real fuel [K] (if known, otherwise set it to 1e+18)
Tbubble_realFuel = 445.00;

%% --- INPUT DATA FOR THE BAYESIAN INFERENCE ANALYSIS --- %%

% Number of components the surrogate mixture will be made of
numComponents = 3;

% Define the hydrocarbon families the candidate surrogate components will be selected from
families = {'nparaffins', 'isoparaffins', 'isoparaffins'};

% Define the range the number of carbon atoms of the hydrocarbon families can vary within
n_ranges = {8:15, 9:12, 13:16};

% Define the range the normalized topochemical atom indices can vary within
LowerBound_eta_B_star = [0 0 0];
UpperBound_eta_B_star = [1 1 1];

% Define the range the mole fractions of each surrogate mixture component can vary within
LowerBound_molFrac = [0.1 0.1 0.1];
UpperBound_molFrac = [0.75 0.75 0.75];

% Construct a cell array made of structure arrays containing detailed information about the whole
% set of candidate species for each surrogate mixture component 
classes = cell(1, numel(families));
for i = 1:numel(families)
    classes{i} = Hydrocarbons(families{i}, n_ranges{i});
end

%% --- IMPORT EXPERIMENTAL DATA --- %%

% Extract the pressure the distillation curve is computed at [Pa]
pressure_distillation = p_distillation(Dataset);

% Import experimental measurements about the thermophysical properties of the real fuel
[fullData, minT_array, maxT_array, variable_names] = data_import(Dataset, families, n_ranges, Tbubble_realFuel, numComponents, pressure_distillation);

%% --- BUILD FUNCTION HANDLES FOR THE PRIOR, LIKELIHOOD, AND POSTERIOR DISTRIBUTIONS --- %%

% Define the (1 x numComponents) vector of concentration parameters for the Dirichlet distribution of molar fractions
alpha = 1*ones(1, numComponents); % unitary values for a symmetric uniform Dirichlet distribution

% Functional handles for the PDFs
[prior, likelihood, posterior] = pdf_build_speed(fullData, families, classes, variable_names, pressure_distillation, LowerBound_molFrac, UpperBound_molFrac, alpha, n_ranges, LowerBound_eta_B_star, UpperBound_eta_B_star);

idx_distillation = find(ismember(variable_names, {'deltaT_dist'}));

fullData_nodist = fullData;
fullData_nodist(idx_distillation) = [];

variables_nodist = variable_names;
variables_nodist(idx_distillation) = [];

% Functional handles for the PDFs
[~, ~, posterior_cheap] = pdf_build_speed_cheap(fullData_nodist, families, classes, variables_nodist, pressure_distillation, LowerBound_molFrac, UpperBound_molFrac, alpha, n_ranges, LowerBound_eta_B_star, UpperBound_eta_B_star);

%% --- DIFFERENTIAL EVOLUTION MARKOV CHAIN (DE-MC) TO EXPLORE POSTERIOR PDF --- %%

% +++ DE-MC parameters +++ %
maxIterations = 5000; % maximum number of iterations for each chain
t_burnin = 500; % number of iterations to be discarded as initial burn-in
t_check = 2*t_burnin; % number of iterations to start R-hat convergence check from
scaling_factor_X = 0.1; % scaling factor for the jump rate concerning molar fractions
scaling_factor_nc = 1.0; % scaling factor for the jump rate concerning numbers of carbon atoms
scaling_factor_eta = 1.0; % scaling factor for the jump rate concerning topochemical atom indices
noise_X = 1e-06; % noise parameter for the surrogate mixture molar fractions
noise_nc = 1e-06; % noise parameter for the surrogate mixture numbers of carbon atoms
noise_eta = 1e-06; % noise par puameter for the surrogate mixture topochemical atom indices
N_chains = 2*(3*numComponents-1); % number of chains
outlier_method = 'mad'; % outlier chain detection method: 'iqr' (interquartile range) or 'mad' (median absolute deviation)
R_hat_threshold = 1.2; % threshold value for the R-hat statistic

% +++ Parallel tempering (PT) parameters +++ %
PT_switch = 'True'; % use parallel tempering ('True') or not ('False')
beta_min = 0.0001; % minimum invterse temperature
T_ladder = 'geometric'; % temperature ladder functional form: 'linear' or 'geometric'
swap_freq = 20; % swap frequency

% +++ Proposal parameters +++ %
n_pairs = 1; % maximum number of chain pairs used to propose the new sample
p_gibbs = 0.1; % probability of a Gibbs move
p_snooker = 0.1; % probability of performing a snooker jump rather than a DE move

[chain, posterior_pdf, chain_reshaped, posterior_pdf_reshaped, AR, R_hat, t_convergence]  = DifferentialEvolutionMarkovChain(families, posterior, posterior_cheap, classes, LowerBound_molFrac, UpperBound_molFrac, n_ranges, LowerBound_eta_B_star, UpperBound_eta_B_star, maxIterations, t_burnin, t_check, scaling_factor_X, scaling_factor_nc, scaling_factor_eta, noise_X, noise_nc, noise_eta, N_chains, outlier_method, R_hat_threshold, PT_switch, beta_min, T_ladder, swap_freq, n_pairs, p_gibbs, p_snooker);

%% --- POST-PROCESSING --- %%

% Return molar fractions, numbers of carbon atoms, and topochemical atom indices characterizing the maximum a posteriori surrogate mixture
[x_MAP, nc_MAP, eta_B_star_MAP] = MAPfinder(posterior_pdf_reshaped, chain_reshaped, numComponents);

% Write a text file and display information about the MAP surrogate
species_MAP = MAPwriter(families, numComponents, x_MAP, nc_MAP, eta_B_star_MAP, fuel_name);

% Visualize Bayesian inference analysis outcomes
band_percentiles = 'True'; % 90% confidence interval colored with percentiles ('True') or not ('False')
confidence_width = 0.95; % 'confidence_width'% confidence interval
[temperature_range, property_MAP, volumeFraction_MAP, percentiles_list, mean_model, percentiles_model, model_output_samples, volFrac_interp, sobol_idx] = props_pushforward(fullData, families, classes, chain, AR, R_hat, chain_reshaped, t_convergence, t_burnin, n_ranges, numComponents, variable_names, x_MAP, nc_MAP, eta_B_star_MAP, minT_array, maxT_array, pressure_distillation, fuel_name, band_percentiles, confidence_width);

%% --- SAVE VARIABLES FROM WORKSPACE --- %%

save('CHRJ_3comp.mat');

File: python/examples/CHRJ_3comp/example_hefa.py
r"""
    ____                  _____ ___    ______
   / __ )____ ___  _____ / ___//   |  / ____/
  / __  / __ `/ / / / _ \\__ \/ /| | / /_
 / /_/ / /_/ / /_/ /  __/__/ / ___ |/ __/
/_____/\\__,_\\__, /\\___/____/_/  |_/_/
            /____/

BayeSAF: Emulation and Design of Sustainable Alternative Fuels
via Bayesian Inference and Descriptors-Based Machine Learning

Contributors / Copyright Notice
© 2026 Jacopo Liberatori — jacopo.liberatori@centralesupelec.fr
Postdoctoral Researcher @ Laboratoire EM2C, CentraleSupélec (CNRS)

© 2026 Davide Cavalieri — davide.cavalieri@uniroma1.it
Postdoctoral Researcher @ Sapienza University of Rome,
Department of Mechanical and Aerospace Engineering (DIMA)

© 2026 Matteo Blandino, Ph.D.

Reference:
J. Liberatori, D. Cavalieri, M. Blandino, M. Valorani, and P.P. Ciottoli.
BayeSAF: Emulation and Design of Sustainable Alternative Fuels via Bayesian
Inference and Descriptors-Based Machine Learning. Fuel 419, 138835 (2026).
Available at: https://doi.org/10.1016/j.fuel.2026.138835.

------------------------------------------------------------------------

Description:
The present example illustrates how to develop a three-component surrogate
mixture emulating the HEFA-SPK POSF-6152 (CHRJ) fuel.

Usage
-----
    cd python
    pip install -e .
    cd examples/CHRJ_3comp
    python example_hefa.py
"""

from __future__ import annotations

import numpy as np

from bayesaf.thermo_transport.hydrocarbons import load_hydrocarbons
from bayesaf.utilities.misc import p_distillation
from bayesaf.utilities.data_import import data_import
from bayesaf.pdfs.pdf_build import build_pdfs, build_pdfs_cheap
from bayesaf.mcmc.demc import run_demc
from bayesaf.postprocessing.map_finder import find_map
from bayesaf.postprocessing.map_writer import write_map
from bayesaf.postprocessing.props_pushforward import props_pushforward
from bayesaf.postprocessing.visualize import visualize_results

# ---------------------------------------------------------------------------
# Real fuel metadata
# ---------------------------------------------------------------------------
DATASET = ["rho", "nu", "distillation", "deltaT_dist",
          "molWeight", "HC", "DCN", "LHV", "flash", "freezing"]
DATA_DIR = "../../../exp/CHRJ_POSF6152"   # directory containing <name>.csv files
FUEL_NAME = "CHRJ POSF-6152"
T_BUBBLE = 445.00       # bubble temperature of the real fuel [K] (if known, otherwise set it to 1e+18)

# ---------------------------------------------------------------------------
# Surrogate configuration
# ---------------------------------------------------------------------------

NUM_COMPONENTS = 3

FAMILIES = ["nparaffins", "isoparaffins", "isoparaffins"]

N_RANGES = [
    list(range(8, 16)),    # nparaffins  C8–C15
    list(range(9, 13)),    # isoparaffins C9–C12
    list(range(13, 17)),    # isoparaffins C13–C16
]

LOWER_ETA = np.zeros(NUM_COMPONENTS)
UPPER_ETA = np.ones(NUM_COMPONENTS)

LOWER_MOL_FRAC = np.array([0.1, 0.1, 0.1])
UPPER_MOL_FRAC = np.array([0.75, 0.75, 0.75])

ALPHA_DIRICHLET = np.ones(NUM_COMPONENTS)   # symmetric uniform Dirichlet

# ---------------------------------------------------------------------------
# DE-MC parameters
# ---------------------------------------------------------------------------

MAX_ITERATIONS   = 5000
T_BURNIN         = 500
T_CHECK          = 2 * T_BURNIN
SCALING_X        = 0.1
SCALING_NC       = 1.0
SCALING_ETA      = 1.0
NOISE_X          = 1e-6
NOISE_NC         = 1e-6
NOISE_ETA        = 1e-6
N_CHAINS         = 2 * (3 * NUM_COMPONENTS - 1)   # 28 chains
OUTLIER_METHOD   = "mad"
R_HAT_THRESHOLD  = 1.2

# Parallel tempering
PT_SWITCH  = "True"
BETA_MIN   = 1e-4
T_LADDER   = "geometric"
SWAP_FREQ  = 20

# Maximum number of chain pairs for proposal
N_PAIRS  = 1

# Proposal probabilities
P_GIBBS    = 0.1
P_SNOOKER  = 0.1

# Post-processing
BAND_PERCENTILES  = True
CONFIDENCE_WIDTH  = 0.95


def main() -> None:
    # ------------------------------------------------------------------
    # 1. Build species databases (classes)
    # ------------------------------------------------------------------
    print("Loading hydrocarbon databases...")
    classes = [load_hydrocarbons(fam, n_ranges) for fam, n_ranges in zip(FAMILIES, N_RANGES)]

    # ------------------------------------------------------------------
    # 2. Import experimental data
    # ------------------------------------------------------------------
    print("Importing experimental data...")
    pressure_distillation = p_distillation(DATASET, data_dir=DATA_DIR)

    full_data, min_T_array, max_T_array, variable_names = data_import(
        DATASET, FAMILIES, N_RANGES, T_BUBBLE, NUM_COMPONENTS, pressure_distillation,
        data_dir=DATA_DIR,
    )

    # ------------------------------------------------------------------
    # 3. Build posterior PDF handles
    # ------------------------------------------------------------------
    print("Building PDF handles...")
    prior_fn, likelihood_fn, posterior_fn = build_pdfs(
        full_data, FAMILIES, classes, variable_names, pressure_distillation,
        LOWER_MOL_FRAC, UPPER_MOL_FRAC, ALPHA_DIRICHLET, N_RANGES,
        LOWER_ETA, UPPER_ETA,
    )

    # Cheap posterior (excludes deltaT_dist)
    idx_dist = [i for i, v in enumerate(variable_names) if v == "deltaT_dist"]
    full_data_nd    = [d for i, d in enumerate(full_data)      if i not in idx_dist]
    variable_names_nd = [v for i, v in enumerate(variable_names) if i not in idx_dist]
    _, _, posterior_cheap_fn = build_pdfs_cheap(
        full_data_nd, FAMILIES, classes, variable_names_nd, pressure_distillation,
        LOWER_MOL_FRAC, UPPER_MOL_FRAC, ALPHA_DIRICHLET, N_RANGES,
        LOWER_ETA, UPPER_ETA,
    )
    
    # ------------------------------------------------------------------
    # 4. Run DE-MC
    # ------------------------------------------------------------------
    print("Running DE-MC...")
    (chain, posterior_pdf, chain_reshaped, posterior_reshaped,
     AR, R_hat, t_convergence) = run_demc(
        families=FAMILIES,
        posterior_fn=posterior_fn,
        posterior_cheap_fn=posterior_cheap_fn,
        classes=classes,
        lower_bound_x=LOWER_MOL_FRAC,
        upper_bound_x=UPPER_MOL_FRAC,
        n_ranges=N_RANGES,
        lower_bound_eta=LOWER_ETA,
        upper_bound_eta=UPPER_ETA,
        max_iterations=MAX_ITERATIONS,
        t_burnin=T_BURNIN,
        t_check=T_CHECK,
        scaling_factor_x=SCALING_X,
        scaling_factor_nc=SCALING_NC,
        scaling_factor_eta=SCALING_ETA,
        noise_x=NOISE_X,
        noise_nc=NOISE_NC,
        noise_eta=NOISE_ETA,
        N_chains=N_CHAINS,
        outlier_method=OUTLIER_METHOD,
        R_hat_threshold=R_HAT_THRESHOLD,
        PT_switch=PT_SWITCH,
        beta_min=BETA_MIN,
        T_ladder=T_LADDER,
        swap_freq=SWAP_FREQ,
        n_pairs=N_PAIRS,
        p_gibbs=P_GIBBS,
        p_snooker=P_SNOOKER,
        log_file="demc_log.txt",
        seed=42,
    )

    converged = t_convergence < MAX_ITERATIONS
    print(f"\nDE-MC {'converged' if converged else 'did NOT converge'} at t = {t_convergence}.")
    
    if R_hat.size > 0:

        rhat_last = R_hat[-1]
    
        # remove NaN entries
        rhat_valid = rhat_last[np.isfinite(rhat_last)]
    
        if rhat_valid.size == 0:
            print("    → R-hat not available (all parameters constant).")
    
        elif np.all(rhat_valid <= R_HAT_THRESHOLD):
            print(f"    → R-hat ≤ {R_HAT_THRESHOLD} for all non-constant parameters: convergence criterion satisfied.")
    
        else:
            print(f"    → R-hat > {R_HAT_THRESHOLD}: chains not yet converged. Increase MAX_ITERATIONS.")
    
    else:
    
        print("  R-hat = not computed (chain shorter than t_check — increase MAX_ITERATIONS)")
        print(f"  mean AR   = {AR[np.isfinite(AR)].mean():.3f} %")

    # ------------------------------------------------------------------
    # 5. MAP estimate
    # ------------------------------------------------------------------
    print("\nFinding MAP surrogate...")
    x_MAP, nc_MAP, eta_B_star_MAP = find_map(posterior_reshaped, chain_reshaped, NUM_COMPONENTS)

    write_map(
        FAMILIES, NUM_COMPONENTS,
        x_MAP, nc_MAP, eta_B_star_MAP,
        FUEL_NAME, output_file="MAP.txt",
    )

    # ------------------------------------------------------------------
    # 6. Pushforward & visualisation
    # ------------------------------------------------------------------
    print("\nComputing property pushforward distributions...")
    pf = props_pushforward(
        full_data, FAMILIES, classes, chain_reshaped,
        NUM_COMPONENTS, variable_names,
        x_MAP, nc_MAP, eta_B_star_MAP,
        min_T_array, max_T_array,
        pressure_distillation,
        confidence_width=CONFIDENCE_WIDTH,
    )

    print("\nGenerating figures...")
    visualize_results(
        full_data, FAMILIES, classes,
        chain, AR, R_hat, chain_reshaped,
        T_BURNIN, R_HAT_THRESHOLD, N_RANGES, NUM_COMPONENTS, variable_names,
        pf,
        FUEL_NAME,
        confidence_width=CONFIDENCE_WIDTH,
        band_percentiles=BAND_PERCENTILES,
        output_dir=".",
    )

    print("\nDone. Results written to MAP.txt and ./Figures/.")


if __name__ == "__main__":
    main()
